Add a main() function to be called from katsdpcontim

KimMcAlpine opened the pull request:
To allow the continuum pipeline to run the QA validation code without calling a separate process, the code is reorganised to be called from a main function.

Additionally Aegean and Bane are now run using the `subprocess` module which allows the validation code to connect to their `stdout` and `stderr` pipes.

mauch commented on Radio_continuum_validation.py:
```diff
@@ -41,9 +41,9 @@
-  -w --write                Write intermediate files generated during processing (e.g. cross-matched
-  and pre-filtered catalogues, etc). This will save having to reprocess the cross-matches, etc when
-  executing the script again. [default: False].
+  -w --write                Write intermedidocoptsate files generated during processing (e.g.
```
intermedidocoptsate -> intermediate

KimMcAlpine replied:
Done

mauch commented on Radio_continuum_validation.py:
```diff
@@ -80,170 +80,212 @@
-        main_dir = script_dir
+def process_args():
+    """Process args to be in the required formats."""
+    cf = currentframe()
+    WARN = '\n\033[91mWARNING: \033[0m' + getframeinfo(cf).filename
```
`WARN` looks like a variable that could be used by any function in the script, so perhaps move these two lines outside the function just after the imports in case you want to use it somewhere else in future.

KimMcAlpine replied:
Done

mauch commented on Radio_continuum_validation.py:
```diff
@@ -80,170 +80,212 @@
+            if args['--fits'] == 'None' or args['--noise'] != 'None':
+                raise SyntaxError
+    except SyntaxError:
+        warnings.warn_explicit("""You must pass in a fits image (option -I) and/or the main catalogue
```
FITS is an acronym and should be capitalised in sentences (as a variable name lowercase is fine). A quick `grep fits` on the script exposes a few other places as well.

Please pardon the pedantry :stuck_out_tongue_winking_eye:

KimMcAlpine replied:
Done

mauch commented on Radio_continuum_validation.py:
```diff
@@ -80,170 +80,212 @@
+        warnings.warn_explicit("""Use option -h to see usage.\n""", UserWarning, WARN, cf.f_lineno)
+        sys.exit()
+
+    # don't use normal display environment unless user wants to view plots on screen
```
Capitalise the first word in a comment ([PEP 8](https://www.python.org/dev/peps/pep-0008/#comments)). This occurs throughout the script.

Pardon the extra pedantry.

KimMcAlpine replied:
Done

mauch commented on filter_config_example.txt:
```diff
@@ -1,7 +1,7 @@
 psf_tol=1.5
-resid_tol=3
\ No newline at end of file
+resid_tol=3
```
Should the file be renamed something like 'filter_config_MeerKAT.txt' since I presume you are setting these parameters specific to the MeerKAT continuum QA?

KimMcAlpine replied:
Done

mauch reviewed (approved):
Looks fine (it seems to be mostly just adding indentation).

Just a couple of minor stylistic issues in the pre-existing code.

While outside the scope of this PR (which is just about getting something working with contim), I reckon that in future it would be nice to:
1. Add `logging` to this script which would mean you only need to handle streaming STDERR and STDOUT to the logger from the Agean and BANE calls and it could be done inside the `execute` function in this package instead of in contim. It would also remove the need for the `WARN` variable here since you can use log.warn.
2. Do a rewrite of `process_args` to use `argparse` rather than `docopts` since then the lengthy docstring at the top of the script can be shortened and also you can parse the necessary arguments directly through to `ArgumentParser.parse_args()` from contim.

I guess we wait for this to be merged to master before ska-sa/katsdpcontim#29 is finalised.

mauch commented:
After starting to look at https://github.com/ska-sa/katsdpcontim/pull/29 I'm wondering if there is a way you can simplify this a little bit by passing the arguments to main rather than using the `redirect_argv` function in contim (which worries me a little bit).

It seems that docopt will take user supplied arguments much like argparse does so if you have this script pass the command line arguments to `main` and the pass those arguments to `process_args` you can call docopt with user supplied arguments.

What I'm thinking is like this:
```
# Pass `sys.argv` to `main` 
if __name__ == "__main__":
    main(sys.argv[1:])

# Then in `main` call `process_args` with `argv`
def main(argv):
    args = process_args(argv)
    .....

# Then in `process_args` call docopt with `argv`
def process_args(argv):
    args = docopt(__doc__, argv=argv)
```

This way you can then pass your own `argv` to main when you call it from katsdpcontim and you preserve the command line arguments when it is run as a script. 
You can then get rid of all of the code that overrides `sys.argv` in contim, and change lines 91--96 of `qa_report.py` in contim from:
```
        command = '/home/kat/valid/Radio_cont_main -I {} --telescope MeerKAT -F'\
                  ' /home/kat/valid/filter_config_example.txt -r'.format(pb_fits)
        sysarg = shlex.split(command)
        with redirect_argv(sysarg):
            with log_qa(log):
                rcv.main()
```

to:
```
        args = ['-I', pb_fits, '--telescope', 'MeerKAT', '-F', '/home/kat/valid/filter_config_MeerKAT.txt', '-r']
        with log_qa(log):
            rcv.main(args)
```

I haven't tested this - and maybe you've tried it already and there is a problem I haven't seen. But this seems neater to me. What do you think?

KimMcAlpine commented:
I agree this is a neater approach, I've implemented the suggested changes.

On Wed, Jan 20, 2021 at 10:00 AM Tom Mauch <notifications@github.com> wrote:

> After starting to look at ska-sa/katsdpcontim#29
> <https://github.com/ska-sa/katsdpcontim/pull/29> I'm wondering if there
> is a way you can simplify this a little bit by passing the arguments to
> main rather than using the redirect_argv function in contim (which
> worries me a little bit).
>
> It seems that docopt will take user supplied arguments much like argparse
> does so if you have this script pass the command line arguments to main
> and the pass those arguments to process_args you can call docopt with
> user supplied arguments.
>
> What I'm thinking is like this:
>
> # Pass `sys.argv` to `main`
> if __name__ == "__main__":
>     main(sys.argv[1:])
>
> # Then in `main` call `process_args` with `argv`
> def main(argv):
>     args = process_args(argv)
>     .....
>
> # Then in `process_args` call docopt with `argv`
> def process_args(argv):
>     args = docopt(__doc__, argv=argv)
>
> This way you can then pass your own argv to main when you call it from
> katsdpcontim and you preserve the command line arguments when it is run as
> a script.
> You can then get rid of all of the code that overrides sys.argv in
> contim, and change lines 91--96 of qa_report.py in contim from:
>
>         command = '/home/kat/valid/Radio_cont_main -I {} --telescope MeerKAT -F'\
>                   ' /home/kat/valid/filter_config_example.txt -r'.format(pb_fits)
>         sysarg = shlex.split(command)
>         with redirect_argv(sysarg):
>             with log_qa(log):
>                 rcv.main()
>
> to:
>
>         args = ['-I', pb_fits, '--telescope', 'MeerKAT', '-F', '/home/kat/valid/filter_config_MeerKAT.txt', '-r']
>         with log_qa(log):
>             rcv.main(args)
>
> I haven't tested this - and maybe you've tried it already and there is a
> problem I haven't seen. But this seems neater to me. What do you think?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ska-sa/MeerKAT-continuum-validation/pull/2#issuecomment-763413410>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACMD2WNJOQDYEFY6SSSUDRDS22ESZANCNFSM4WG6LPUQ>
> .
>
